<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#-Комментарий-ревьюера" data-toc-modified-id="-Комментарий-ревьюера-5"><span class="toc-item-num">5&nbsp;&nbsp;</span> Комментарий ревьюера</a></span></li></ul></div>

# Защита персональных данных клиентов

Нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# заранее установим в константу random_state
random_state = 42

In [2]:
df = pd.read_csv('/datasets/insurance.csv')

In [3]:
# функция вывода информации о датасете
def data_info(data):
    display('общая информация')
    display(data.head())
    display('_______________________________________________________________')
    display('информация о типах данныx')
    display()
    display(data.info())
    display('_______________________________________________________________')
    display('описание')
    display()
    display(data.describe())
    display('_______________________________________________________________')
    display('пропуски')
    display(df.isna().sum())
    display()
    display('_______________________________________________________________')
    display('дубликаты')
    display()
    display(data.duplicated().sum())
    display('_______________________________________________________________')
    display('корреляция')
    display()
    display(pd.DataFrame(data.corr()))

In [4]:
data_info(df)

'общая информация'

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

'_______________________________________________________________'

'информация о типах данныx'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

'_______________________________________________________________'

'описание'

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

'_______________________________________________________________'

'пропуски'

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

'_______________________________________________________________'

'дубликаты'

153

'_______________________________________________________________'

'корреляция'

Пол   Возраст  Зарплата  Члены семьи  \
Пол                1.000000  0.002074  0.014910    -0.008991   
Возраст            0.002074  1.000000 -0.019093    -0.006692   
Зарплата           0.014910 -0.019093  1.000000    -0.030296   
Члены семьи       -0.008991 -0.006692 -0.030296     1.000000   
Страховые выплаты  0.010140  0.651030 -0.014963    -0.036290   

                   Страховые выплаты  
Пол                         0.010140  
Возраст                     0.651030  
Зарплата                   -0.014963  
Члены семьи                -0.036290  
Страховые выплаты           1.000000

переменуем признаки для корректного анализа и обработки данных

In [5]:
df.columns = ['sex', 'age', 'salary', 'family_members', 'insurance_payments']

**Вывод:**

1. Данные представлены в датасете df


2. df содержит 5000 объектов и 5 признаков 

    - 'Пол' - int64 
    
    - 'Возраст' - float64
    
    - 'Зарплата' - float64
    
    - 'Члены семьи' - int64
    
    - 'Страховые выплаты' - int64
    

3. Пропуски отсутствуют


4. Присутствуют полные дубликаты в количестве 153.


5. Присутствует заметная линейная связь 0.65 признаков 'Страховые выплаты' и 'Возраст'

анализируем дубликаты - не понятно, почему метод duplicated() выдает 153 объекта, по факту объекты не выглядят как дубликаты???

In [6]:
df_dublicate = df[df.duplicated()]

In [7]:
df_dublicate

sex   age   salary  family_members  insurance_payments
281     1  39.0  48100.0               1                   0
488     1  24.0  32900.0               1                   0
513     0  31.0  37400.0               2                   0
718     1  22.0  32600.0               1                   0
785     0  20.0  35800.0               0                   0
...   ...   ...      ...             ...                 ...
4793    1  24.0  37800.0               0                   0
4902    1  35.0  38700.0               1                   0
4935    1  19.0  32700.0               0                   0
4945    1  21.0  45800.0               0                   0
4965    0  22.0  40100.0               1                   0

[153 rows x 5 columns]

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Если $P$ - обратимая квадратная матрица $n$ на $m$, то преобразованная матрица это произведение $X$ на $P$. 

Получаем предсказание преобразованной матрицы

$$
a_1 = XPw_1
$$

Здесь $XP$ прямоугольная матрица $m$ на $n$, а формула обучения:
$$
w_1 = ((XP)^T\cdot XP)^{-1}\cdot (XP)^T y
$$

Требуется доказать равенство 

$$
a = a_1
$$

**Доказательство:**

Предсказания для исходной матрицы

$$
a = Xw = X(X^T X)^{-1} X^T y 
$$

Предсказания для преобразованной матрицы $a_1$:

$$
a_1 = XP \cdot w_1 = XP \cdot ((XP)^T\cdot XP)^{-1}\cdot (XP)^T y  
$$

Раскроем транспонирование:
$$
a_1 = XP \cdot (P^TX^TXP)^{-1}\cdot P^TX^Ty = XP \cdot (P^T(X^TX)P)^{-1}\cdot P^TX^Ty 
$$

Три матрицы под (-1) квадратные и обратимые, поэтому:
$$
a_1 = XP \cdot P^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty = XE(X^TX)^{-1}EX^Ty = X(X^TX)^{-1}X^Ty = a
$$

Т.о. доказано, что предсказания не изменяются при умножении исходной матрицы признаков на произвольную обратимую матрицу.

**Ответ:** при умножении признаков на обратимую матрицу качество линейной регрессии не изменяется

**Обоснование:** равенство $a = a_1$ доказывает, что примененный алгоритм преобразования матрицы не ухудшает качество моделей машинного обучения

## Алгоритм преобразования

**Алгоритм**

1. Генерируем рандомную матрицу
2. Проверяем на обратимость 
3. Умножаем на матрицу признаков
4. Обучаем модель 
5. Определяем метрику R2 с помощью линейной регрессии для предобразованных данных и исходных, а затем сравниваем

**Обоснование**

Качество линейной регрессии не поменяется, это продемонстрировало равенство
$$
wP = P ^{-1}w
$$

## Проверка алгоритма

Разделим признаки и таргет и приступим к теоретической части.

In [8]:
features = df.drop('insurance_payments', axis = 1)
target = df['insurance_payments']

Функция для алгоритма преобразования. Проверять определеитель будет с точностью до эпсилон.

функция принимает матрицу признаков и точность эпсилон
функция выдает преобразованную матрицу признаков

In [9]:
def transform(features, eps):
    #m = features.shape[0]
    n = features.shape[1]
    det = 0
    while abs(det) < eps:
        P = np.random.randint(-50, 50, size=(n, n))
        det = np.linalg.det(P)
        print('det =', det)
    return features.dot(P)

In [10]:
eps = 10**(-4)
features_P = transform(features, eps=eps)

det = -125639.0000000001


In [11]:
features_P.head()

0         1          2         3
0  1389479.0 -446567.0  1685936.0 -348113.0
1  1064722.0 -342174.0  1291508.0 -266983.0
2   588435.0 -189087.0   713681.0 -147638.0
3  1167979.0 -375435.0  1417597.0 -292304.0
4   731252.0 -234992.0   887065.0 -183356.0

Получена случайная обратимая матрица - ненулевой определитель. 
Умножением исходной матрицы на полученную получаем преобразованную.

Для проверки алгоритма обучим линейную регрессию на исходной и преобразованной матрицах, сравним R2.

Выделяем валидационную выборку

In [12]:
features_train, features_valid, target_train, target_valid = train_test_split(features, target,  
                                                                              test_size=0.25, random_state=random_state)
features_P_train, features_P_valid = train_test_split(features_P, test_size=0.25, random_state=random_state)

Обучение на исходной матрице

In [13]:
model1 = LinearRegression()
model1.fit(features_train, target_train)
score1 = r2_score(target_valid, model1.predict(features_valid))
print('R2 на исходной матрице:', score1)

R2 на исходной матрице: 0.4254778540696319


Обучение на преобразованной матрице

In [14]:
model2 = LinearRegression()
model2.fit(features_P_train, target_train)
score2 = r2_score(target_valid, model2.predict(features_P_valid))
print('R2 на преобразованной матрице:', score2)

R2 на преобразованной матрице: 0.4254778540699664


Сравниваем R2

In [15]:
print('Разница метрик:', abs(score1 - score2))

Разница метрик: 3.3451019731955967e-13


**Вывод:**

Решена задача защиты данных клиентов страховой компании «Хоть потоп». Разработан метод преобразования данных, после которого персональная информация защищена. Задача реализована пошаговым алгоритмом. Проверено, что качество моделей на исходных данных и преобразованных не отличается.